### PR TITLE
Reorder render sets, refactor bevy_sprite to take advantage

### PR DESCRIPTION
**Pull request by james-j-obrien**
This is a continuation of this PR: #8062 

# Objective

- Reorder render schedule sets to allow data preparation when phase item order is known to support improved batching
- Part of the batching/instancing etc plan from here: https://github.com/bevyengine/bevy/issues/89#issuecomment-1379249074
- The original idea came from @inodentry and proved to be a good one. Thanks!
- Refactor `bevy_sprite` and `bevy_ui` to take advantage of the new ordering

## Solution
- Move `Prepare` and `PrepareFlush` after `PhaseSortFlush` 
- Add a `PrepareAssets` set that runs in parallel with other systems and sets in the render schedule.
  - Put prepare_assets systems in the `PrepareAssets` set
  - If explicit dependencies are needed on Mesh or Material RenderAssets then depend on the appropriate system.
- Add `ManageViews` and `ManageViewsFlush` sets between `ExtractCommands` and Queue
- Move `queue_mesh*_bind_group` to the Prepare stage
  - Rename them to `prepare_`
  - Put systems that prepare resources (buffers, textures, etc.) into a `PrepareResources` set inside `Prepare`
  - Put the `prepare_..._bind_group` systems into a `PrepareBindGroup` set after `PrepareResources`
- Move `prepare_lights` to the `ManageViews` set
  - `prepare_lights` creates views and this must happen before `Queue`
  - This system needs refactoring to stop handling all responsibilities
  - Gather lights, sort, and create shadow map views. Store sorted light entities in a resource

- Remove `BatchedPhaseItem`
- Replace `batch_range` with `batch_size` representing how many items to skip after rendering the item or to skip the item entirely if `batch_size` is 0.
- `queue_sprites` has been split into `queue_sprites` for queueing phase items and `prepare_sprites` for batching after the `PhaseSort`
  - `PhaseItem`s are still inserted in `queue_sprites`
  - After sorting adjacent compatible sprite phase items are accumulated into `SpriteBatch` components on the first entity of each batch, containing a range of vertex indices. The associated `PhaseItem`'s `batch_size` is updated appropriately.
  - `SpriteBatch` items are then drawn skipping over the other items in the batch based on the value in `batch_size`
- A very similar refactor was performed on `bevy_ui`
---

## Changelog

Changed:
- Reordered and reworked render app schedule sets. The main change is that data is extracted, queued, sorted, and then prepared when the order of data is known.
- Refactor `bevy_sprite` and `bevy_ui` to take advantage of the reordering.

## Migration Guide
- Assets such as materials and meshes should now be created in `PrepareAssets` e.g. `prepare_assets<Mesh>`
- Queueing entities to `RenderPhase`s continues to be done in `Queue` e.g. `queue_sprites`
- Preparing resources (textures, buffers, etc.) should now be done in `PrepareResources`, e.g. `prepare_prepass_textures`, `prepare_mesh_uniforms`
- Prepare bind groups should now be done in `PrepareBindGroups` e.g. `prepare_mesh_bind_group`
- Any batching or instancing can now be done in `Prepare` where the order of the phase items is known e.g. `prepare_sprites`

 
## Next Steps
 - Introduce some generic mechanism to ensure items that can be batched are grouped in the phase item order, currently you could easily have `[sprite at z 0, mesh at z 0, sprite at z 0]` preventing batching.
 - Investigate improved orderings for building the MeshUniform buffer
 - Implementing batching across the rest of bevy

### CODE REVIEW

**Comment by james-j-obrien** [comment]
Had to add an additional check in `prepare_sprites` to ensure that the image asset is prepared. This isn't ideal as we then do the lookup again in `queue_sprites` in order to create the bind group.

File: crates/bevy_render/src/lib.rs
Comment (JMS55) [inline crates/bevy_render/src/lib.rs:110]:
```suggestion
    /// Create [`BindGroups`](crate::render_resource::BindGroup) that depend on those resources.
```

File: crates/bevy_render/src/render_phase/mod.rs
Comment (JMS55) [inline crates/bevy_render/src/render_phase/mod.rs:94]: We can't remove render_range(). It's important to this PR https://github.com/bevyengine/bevy/pull/8015.

File: crates/bevy_render/src/render_phase/mod.rs
Comment (JMS55) [inline crates/bevy_render/src/render_phase/mod.rs:126]: Wording of "if the same type is batched" is awkward.

**Review by JMS55 — COMMENTED** [review]
Generally looks good to me, but I can really only review bevy_pbr/bevy_render. I'm not familiar with bevy_sprite, bevy_ui, etc.

Also, curious why the QueueMeshes subset exists.

File: crates/bevy_render/src/render_phase/mod.rs
Comment (james-j-obrien) [reply crates/bevy_render/src/render_phase/mod.rs]: I've improved the wording here, let me know if it sounds reasonable.

Comment (JMS55) [reply crates/bevy_render/src/render_phase/mod.rs]: Looks good now.

File: crates/bevy_ui/src/render/mod.rs
Comment (superdump) [inline crates/bevy_ui/src/render/mod.rs:622]: Why is the batching being done in queue, before sorting? What if other plugins queue things to the TransparentUi render phase? That would break things.

**Review by superdump — CHANGES_REQUESTED** [review]
Mostly looks OK to me. The approach with UI to do batching in Queue looks at first glance like it would break if other plugins queued anything to the TransparentUi render phase such that those items should be sorted between.

Other than that I'd like to see some benchmarks and either decide that performance is good enough that more performance improvements can be made in a follow-up PR to save on blocking merge of this PR, or that more work is needed.

**Comment by superdump** [comment]
On my M1 Max, this PR drops `bevymark -- 10000 16` performance from ~72fps to ~44fps, so I think we will have to do some work to optimise this before merging as it's too far away. I'll have a look at what I can do to help this along. :)

File: crates/bevy_sprite/src/render/mod.rs
Comment (james-j-obrien) [reply crates/bevy_sprite/src/render/mod.rs:659]: The end for each of these is updated after the vertices are added to the buffers so the end value isn't important. This change would make the type RangeInclusive instead of Range so not doing this immediately but if there's a reason I'm not seeing am open to suggestions.

**Comment by james-j-obrien** [comment]
Fixed the UI implementation to match that of `bevy_sprite`. Main caveat here is that I have set the z depth used for the phase sort in `TransparentUi` to use the `stack_index` of the corresponding ui node as an f32. This works and still allows for positioning arbitrary non node `TransparentUi` items behind or in front of the UI, but the value is also arbitrary so open to suggestions if there is a cleaner method I'm not seeing. One alternative is to sort by the usize `stack_index` instead of using a float at all but that would then also apply to non node `TransparentUi` items for little upside.

File: crates/bevy_sprite/src/render/mod.rs
Comment (superdump) [reply crates/bevy_sprite/src/render/mod.rs:659]: Nope. Got it.

**Comment by nicopap** [comment]
There is currently a lot of work involved in improving UI perfs. See the implementation by @ickshonpe and subsequent discussion:

* https://github.com/bevyengine/bevy/pull/9212

File: crates/bevy_render/src/gpu_component_array_buffer.rs
Comment (superdump) [inline crates/bevy_render/src/gpu_component_array_buffer.rs:23]: This is not what is intended. Per-frame dynamic data is meant to still be prepared in prepare.

File: crates/bevy_render/src/render_phase/mod.rs
Comment (superdump) [inline crates/bevy_render/src/render_phase/mod.rs:89]: This feels like it would be significantly slower than maintaining an index variable and skipping over large numbers of items at once based on the batch size.

File: crates/bevy_sprite/src/render/mod.rs
Comment (superdump) [inline crates/bevy_sprite/src/render/mod.rs:475]: Is there any noticeable performance impact of using a full `Handle<Image>` instead of `HandleId`?

File: crates/bevy_sprite/src/render/mod.rs
Comment (superdump) [inline crates/bevy_sprite/src/render/mod.rs:717]:
```suggestion
                    // Override the size if a custom one is specified
```

**Review by superdump — CHANGES_REQUESTED** [review]

File: crates/bevy_render/src/gpu_component_array_buffer.rs
Comment (james-j-obrien) [reply crates/bevy_render/src/gpu_component_array_buffer.rs:23]: You're right that this shouldn't be in PrepareAssets (doesn't make any sense) but the GpuArrayBufferIndex is currently consumed in queue_material_meshes so either we need to break that dependency so we can place this system in prepare or find another place to put it.

File: crates/bevy_sprite/src/render/mod.rs
Comment (james-j-obrien) [reply crates/bevy_sprite/src/render/mod.rs:475]: I don't think there would be since it's only one per batch and the size difference is small, but I also don't recall why I made this change. I believe I noticed that in UiBatch we were using a handle instead of an id, tried swapping to an id for consistency and then swapped back for some reason that I now can't remember. Ideally since HandleId is smaller the extracted ui nodes should be using ids as well. I will try and swap them all to ids and see if I rediscover why I made this change.

File: crates/bevy_render/src/gpu_component_array_buffer.rs
Comment (james-j-obrien) [reply crates/bevy_render/src/gpu_component_array_buffer.rs:23]: The reason that it is used in queue_material_meshes is so that it can be used in the sort key.

File: crates/bevy_sprite/src/render/mod.rs
Comment (james-j-obrien) [reply crates/bevy_sprite/src/render/mod.rs:475]: Ok so the comment attached to the image_handle_id field on ExtractedSprite says:
```
PERF: storing a `HandleId` instead of `Handle<Image>` enables some optimizations (`ExtractedSprite` becomes `Copy` and doesn't need to be dropped)
```
We don't need ExtractedSprite to be Copy anymore so the comment here no longer applies, but I have swapped back to ids since they should be smaller

File: crates/bevy_render/src/gpu_component_array_buffer.rs
Comment (superdump) [reply crates/bevy_render/src/gpu_component_array_buffer.rs:23]: Right, this comes to the key point of the PR. I had implemented this before as queuing with a default dynamic offset value (I think maybe None) and then filling it in in prepare, where batching will be done for material meshes. Now that we have `GpuArrayBuffer` and the `GpuComponentArrayBufferPlugin` is being used for `MeshUniform`, this code path changed. I think we will have to stop using `GpuComponentArrayBufferPlugin` for `MeshUniform` and instead make a custom prepare_mesh_uniforms system that does what the plugin does, but also updates the dynamic offset of the corresponding phase item.

Comment (james-j-obrien) [reply crates/bevy_render/src/gpu_component_array_buffer.rs]: Implemented outside the plugin in `prepare_mesh_uniform_buffer`

File: crates/bevy_pbr/src/render/mesh.rs
Comment (JMS55) [inline crates/bevy_pbr/src/render/mesh.rs:118]: Why was this removed?

**Comment by JMS55** [comment]
Missing adapting TAA and I think some other misc things like the game of life compute shader example.

File: crates/bevy_pbr/src/render/mesh.rs
Comment (james-j-obrien) [reply crates/bevy_pbr/src/render/mesh.rs:118]: Building the MeshUniform buffer is now done in `prepare_mesh_uniform_buffer` so it can be done in sorted order.

**Comment by james-j-obrien** [comment]
> Missing adapting TAA and I think some other misc things like the game of life compute shader example.

I believe the `computer_shader_game_of_life.rs` has been updated to move bind group creation to prepare and it works when I run it locally is there something specific you are referring to?

TAA also works as expected, perhaps moving the texture creation to ManageViews would be more consistent but there's no dependency that requires them to be created before Prepare. Happy to make that change though.

**Comment by JMS55** [comment]
Sorry I somehow missed that you did update game of life.

Yes I think it's better to move TAA to be consistent :)

**Comment by robtfm** [comment]
Render schedule graph (i haven't looked in detail yet)
[rendergraph.dot.zip](https://github.com/bevyengine/bevy/files/12386294/rendergraph.dot.zip)

**Comment by superdump** [comment]
And here's an SVG of it:
rendergraph

**Comment by robtfm** [comment]
I’ve started going through. There are some unneeded apply_deferreds, some odd nested schedules, some ambiguous resources and generally I think it needs a fair bit of work (though I think a number of the problems are present in main as well). 

It’s hard to work out the best way to even think about it. I think I’m settling on using a “per resource” view to trace each bit of data through the stages and systems, but it’s very manual. Hope to have at least a view on what we can/should change later today. And might try to play with the debugdump crate to help with generating these kind of data-flow oriented views as well

**Comment by superdump** [comment]
That would be fantastic! We should also limit this PR to change only the simple things. Things like reworking prepare_lights in a proper way should wait, in my opinion. Otherwise the PR will be changing too many things at once.

**Comment by robtfm** [comment]
[render schedule order.xlsx](https://github.com/bevyengine/bevy/files/12397011/render.schedule.order.xlsx)

attached shows the resource/component usage of all the systems, roughly in execution order (though the order is ambiguous, the order i put is valid). 
it's manually created so may have bugs. i can't see how to generate this currently unfortunately. also i noticed that some systems in it are missing from the graph above (prepare_ssao_* e.g.), not sure why that is.
I ignored components/resources that are inserted/modified in `Extract`, and those that are only used in the `Render` stage.

the red cells are those with dependency issues currently. these are mainly one-frame-lags, some just for initial creation (e.g. prepare_mesh_bind_group vs prepare_mesh_uniforms), some ongoing (prepare_taa_jitter_and_mip_bias vs prepare_ssao_pipelines), and some potentially crashing (prepare_uniform_components::<BloomUniforms> vs prepare_bloom_bind_groups).

i think we should have 
- a `PrepareUniforms` or `PrepareBuffers` stage which covers about half of the current `Prepare` stage
- `PrepareBindgroups` after `PrepareUniforms/Buffers` (and after flush due to ViewUniforms) which covers the rest
- non-dependent prepares can float free (e.g. most pipelines)

- prepare_ssao_textures, prepare_bloom_textures, prepare_taa_history_textures and prepare_prepass_textures should be in `PrepareUniforms/PrepareBuffers`

- prepare_materials and prepare_materials_2d should be in `PrepareAssets`, queue_materials, queue_materials_2d, queue_shadows and queue_prepass_material_meshes should be `.after()` them

- the `PrepareAsset` substages (`PreAssetPrepare`, `AssetPrepare`, `PostAssetPrepare`) don't server much value. systems should be `.after(prepare_assets::<C>)` for whatever they need. there are several deps missing like this (queue_wireframes.after(prepare_assets::<Mesh>, e.g.)

- prepare_component_uniforms::<C> should be in `PrepareUniforms/PrepareBuffers`

- prepare_fog should be in `PrepareUniforms/PrepareBuffers`

then, 
- `QueueFlush` and `PhaseSortFlush` are not needed
- only 1 apply_deferred is needed in Prepare, in between PrepareUniforms and PrepareBindgroups. currently there are 4, and none of them actually hit the insert in prepare_uniforms.

i'll try to make a second tab with these suggestions as well for clarity in a bit.

File: crates/bevy_pbr/src/render/mesh.rs
Comment (robtfm) [inline crates/bevy_pbr/src/render/mesh.rs:409]: seen is never modified here

**Review by robtfm — COMMENTED** [review]
also one specific comment on the new code below, and a handful of unrelated things i came across while digging through (for separate issues), just dropping them here to not lose them

- prepare_mesh2d_view_bind_groups - the group is generated regardless of any phase items. also the bindgroup is identical across views but a bindgroup is created per view anyway

- prepare_lights makes a shadow view per shadowmap per view, i didn't look in depth but this seems unnecessary.

- prepare_view_tonemapping_pipelines - why `With<ViewTargets>`? this stops tonemapping being used on intermediate cameras.

**Comment by superdump** [comment]
I agree with splitting `Prepare` into `PrepareBuffers` and `PrepareBindGroups`. This is basically what Jasmine was suggesting. Separately, as mentioned in DMs, I was thinking about whether further splitting would be desirable in order to make use of buffers more flexible (allow others to push into the main buffers like MeshUniform) and to maybe have a way to do something useful while waiting for `write_buffer` to complete. My thinking there was that there are the following pieces of work:
* Preparation of data into system RAM.
  * This could be done flexibly by multiple systems in something like a `PrepareBuffers` or `PrepareData` set.
  * Note that currently prepare systems sometimes push data into temporary Vecs before they are serialised by encase into the buffer that is passed to wgpu. Preparation of data should probably complete the serialisation step too.
* Allocation of VRAM buffers into which the data can be copied.
  * Once all systems have run to prepare data in system RAM, required VRAM buffer capacities are known and so they can be allocated.
  * This isn't needed every frame, but if it is needed, then the `wgpu::Buffer` changes and `BindGroup`s have to be (re)created. After allocation has been done, bind groups can be created.
* Submission of commands to copy data from system RAM to VRAM
  * This could be done on a separate thread so that systems aren't blocked while waiting for the VRAM buffers to become available to copy into.
* (Re)creation of `BindGroup`s
  * Just has to be after allocation.
* Assuming there are no locks in wgpu that would block us, we could start encoding commands in parallel with the wait for buffer data to be updated. The only thing we can't do is drop command buffers until the buffers in bind groups on which they depend have been updated.

Ideally for performance, systems define their explicit dependencies. It's just that that reduces ergonomics if you need to have a sync between the dependencies and dependents.

As for `prepare_lights`, I think all lights could be global, aside from whether they are relevant to a view (e.g. point lights are where they are and could be visible by multiple views in the same area, but views in completely different regions of the scene might have completely disjoint sets of lights.) Point/spot light shadow maps are reusable across views. Directional light shadow maps are per-view at least with our current setup. Directional light shadow maps need one texture view per cascade. Point light shadow maps need 6, one for each cubemap face. Spot lights need one.

**Comment by robtfm** [comment]
s/s with suggested order : [render schedule order.xlsx](https://github.com/bevyengine/bevy/files/12402112/render.schedule.order.xlsx)

if everyone is comfortable i'm happy to make a pr to @james-j-obrien 's branch

**Comment by JMS55** [comment]
Can we rename PrepareBuffers to PrepareResources or something more generic?

File: crates/bevy_pbr/src/render/mesh.rs
Comment (superdump) [reply crates/bevy_pbr/src/render/mesh.rs:405]: Looks like a mistake. seen should probably be updated with the mesh index inside the if block.

Comment (james-j-obrien) [reply crates/bevy_pbr/src/render/mesh.rs]: Yeah whoops, fixed and moved the FixedBitSet to a Local like we do for visible_entities.

**Review by superdump — CHANGES_REQUESTED** [review]
Looking pretty good to me. Huge thanks @robtfm for going through all the systems and sets!

File: crates/bevy_pbr/src/ssao/mod.rs
Comment (robtfm) [reply crates/bevy_pbr/src/ssao/mod.rs:120]: This (and the other pipelines) are just in `Prepare` because they don't have any dependencies (before or after) with any other `Prepare` systems. All the ones in `PrepareResource` are in there because something in `PrepareBindgroups` relies on them being done first.

File: crates/bevy_render/src/mesh/mod.rs
Comment (robtfm) [reply crates/bevy_render/src/mesh/mod.rs:23]:
```suggestion
            // 'Mesh' must be prepared after 'Image' as meshes rely on the morph target image being ready
            .add_plugins(RenderAssetPlugin::<Mesh, Image>::default());
```

**Comment by superdump** [comment]
The new SVG is looking good! :) It feels like it makes more sense.

File: crates/bevy_render/src/render_asset.rs
Comment (robtfm) [reply crates/bevy_render/src/render_asset.rs:49]:
```suggestion
/// [`RenderSet::PrepareAssets`](crate::RenderSet::PrepareAssets) steps for the specified [`RenderAsset`].
/// 
/// The `AFTER` generic parameter can be used to specify that `A::prepare_asset` should not be run until
/// `prepare_assets::<AFTER>` has completed. This allows the prepare_asset function to depend on another 
/// prepared RenderAsset, for example `Mesh::prepare_asset` relies on `RenderAssets::<Image>` for morph
/// targets, so the plugin is created as `RenderAssetPlugin::<Mesh, Image>::default()`.
pub struct RenderAssetPlugin<A: RenderAsset, AFTER: RenderAssetDependency + 'static = ()> {
```

File: crates/bevy_render/src/lib.rs
Comment (robtfm) [inline crates/bevy_render/src/lib.rs:163]:
```suggestion
       schedule.configure_sets((ExtractCommands, PrepareAssets, Prepare).chain());
```
it's not strictly necessary at the moment, but i think it would be useful to add an extra constraint to force `PrepareAssets` to finish before `Prepare`. Lots of the `Prepare` systems use `RenderAssets<Image>` and this will avoid breaking if we remove the need for images from earlier systems. Also without this there's no constraint from `PrepareAssets` to `Render` so if people use custom renderassets for the first time in the rendergraph it could theoretically blow up.

File: crates/bevy_render/src/view/window/mod.rs
Comment (robtfm) [inline crates/bevy_render/src/view/window/mod.rs:45]: i removed all the other stage wrappers but missed this one, could we please remove `WindowSystem::Prepare` as well (and make prepare_view_targets `.after(prepare_windows)` directly)?

**Review by robtfm — APPROVED** [review]
looks good to me. i tested a number of examples and didn't see any issues. couple of minor points but happy to approve as is.

File: crates/bevy_render/src/lib.rs
Comment (james-j-obrien) [reply crates/bevy_render/src/lib.rs:163]: I totally agree here, looking at the render graph the main oddity I noticed was the lack of constraints on PrepareAssets relative to the rest of the graph.

File: crates/bevy_render/src/render_phase/mod.rs
Comment (JMS55) [inline crates/bevy_render/src/render_phase/mod.rs:93]: Curious what the assembly looks like for this block.

**Review by JMS55 — APPROVED** [review]
An absolutely amazing and foundational PR :). This resolves my long standing complaint of "why the heck are we making bind groups in a stage (system set? is that the official term now?) called queue?". Thanks so much for doing this!

File: crates/bevy_render/src/render_phase/mod.rs
Comment (james-j-obrien) [reply crates/bevy_render/src/render_phase/mod.rs:93]: Here's some annotated output from just the main render loop: [render.txt](https://github.com/bevyengine/bevy/files/12414752/render.txt)
And here's the full asm for the function: [render_full.txt](https://github.com/bevyengine/bevy/files/12414757/render_full.txt)
(had to use txt files as github complains about .asm)

Comment (JMS55) [reply crates/bevy_render/src/render_phase/mod.rs:93]: Thanks. Don't think it's a huge problem then. Branch predictor probably handles it fine.

File: crates/bevy_core_pipeline/src/core_3d/mod.rs
Comment (superdump) [reply crates/bevy_core_pipeline/src/core_3d/mod.rs:152]: I'll do that separately. :)

**Review by superdump — APPROVED** [review]
@james-j-obrien - let's get this merged! Please update to main and ping me on Discord so I can click the button. :)

**Pull request merged**
